flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/Observer.java

```diff
@@ -0,0 +1,182 @@
+import org.apache.flink.kubernetes.operator.crd.status.FlinkDeploymentStatus;
+import org.apache.flink.kubernetes.operator.crd.status.JobStatus;
+import org.apache.flink.kubernetes.operator.service.FlinkService;
+import org.apache.flink.runtime.client.JobStatusMessage;
+
+import io.fabric8.kubernetes.api.model.apps.Deployment;
+import io.fabric8.kubernetes.api.model.apps.DeploymentSpec;
+import io.fabric8.kubernetes.api.model.apps.DeploymentStatus;
+import io.javaoperatorsdk.operator.api.reconciler.Context;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+/** Observes the actual state of the running jobs on the Flink cluster. */
+public class Observer {
+
+    private static final Logger LOG = LoggerFactory.getLogger(Observer.class);
+
+    private final FlinkService flinkService;
+
+    public Observer(FlinkService flinkService) {
+        this.flinkService = flinkService;
+    }
+
+    public boolean observe(
+            FlinkDeployment flinkApp, Context context, Configuration effectiveConfig) {
+        observeJmDeployment(flinkApp, context, effectiveConfig);
+        return isReadyToReconcile(flinkApp, effectiveConfig);
+    }
+
+    private void observeJmDeployment(
+            FlinkDeployment flinkApp, Context context, Configuration effectiveConfig) {
+        FlinkDeploymentStatus deploymentStatus = flinkApp.getStatus();
+        JobManagerDeploymentStatus previousJmStatus =
+                deploymentStatus.getJobManagerDeploymentStatus();
+
+        if (JobManagerDeploymentStatus.READY == previousJmStatus) {
+            return;
+        }
+
+        if (JobManagerDeploymentStatus.DEPLOYED_NOT_READY == previousJmStatus) {
+            deploymentStatus.setJobManagerDeploymentStatus(JobManagerDeploymentStatus.READY);
+            return;
+        }
+
+        Optional<Deployment> deployment = context.getSecondaryResource(Deployment.class);
+        if (deployment.isPresent()) {
+            DeploymentStatus status = deployment.get().getStatus();
+            DeploymentSpec spec = deployment.get().getSpec();
+            if (status != null
+                    && status.getAvailableReplicas() != null
+                    && spec.getReplicas().intValue() == status.getReplicas()
+                    && spec.getReplicas().intValue() == status.getAvailableReplicas()) {
+
+                if (flinkService.isJobManagerPortReady(effectiveConfig)) {
+                    // typically it takes a few seconds for the REST server to be ready
+                    LOG.info(
+                            "JobManager deployment {} in namespace {} port ready, waiting for the REST API...",
+                            flinkApp.getMetadata().getName(),
+                            flinkApp.getMetadata().getNamespace());
+                    deploymentStatus.setJobManagerDeploymentStatus(
```

Comment:
What will happen if all the replicas are running, but the rest port is not reachable? We will have the `MISSING`, not the `DEPLOYED_NOT_READY`.

Reply:
Exactly, thats how it works and this is also how it worked before, it stays in DEPLOYING until the replicas and ports are ready when it goes to DEPLOYED_NOT_READY. The name might not be the best but it means that we need to do one more iteration because the rest api probably need a few more seconds (which we cannot explicitly check)

Reply:
This is now highlighted in the observer test: https://github.com/apache/flink-kubernetes-operator/pull/26/files#diff-ac210d80370718839b3bdfc124a9dfc2ed32b06e35ca71474efd5a9f27203595R92-R116

Reply:
You are right.